src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -1,46 +1,100 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Options.Providers;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Options
 {
     [Export(typeof(IGlobalOptionService)), Shared]
     internal class GlobalOptionService : IGlobalOptionService
     {
-        private readonly Lazy<HashSet<IOption>> _options;
+        private readonly Lazy<ImmutableHashSet<IOption>> _lazyAllOptions;
         private readonly ImmutableArray<Lazy<IOptionPersister>> _optionSerializers;
+        private readonly ImmutableDictionary<string, ImmutableArray<Lazy<IOptionProvider, LanguageMetadata>>> _languageToLazyProvidersMap;
+        private readonly Dictionary<string, ImmutableHashSet<IOption>> _languageToSerializableOptionsMap;
```

Comment:
Since `_languageToLazyProvidersMap` is only used to calculate `_languageToSerializableOptionsMap` wouldn't it make sense to have a single dictionary `ImmutableDictionary<string, Lazy<ImmutableHashSet<IOption>>>`, where the Lazy constructor lambda captures the lazy option providers (`ImmutableArray<Lazy<IOptionProvider, LanguageMetadata>>`) for the language?

Reply:
Good idea, done.